taar/recommenders/similarity_recommender.py

```diff
@@ -0,0 +1,130 @@
+    return utils.get_s3_json_content(s3_bucket, DONOR_LIST_KEY)
+
+
+def fetch_lr_curve_remote_resource():
+    return utils.get_s3_json_content(s3_bucket, SIM_TO_P_FUN_KEY)
+
+
+class SimilarityRecommender(BaseRecommender):
+    """ A recommender class that returns top N addons based on the client similarity
+    with a set of candidate addon donors. Several telemetry fields are sued to compute
+    pairwise similarity with the donors and similarities are converted into a likelihood
+    ratio of being a good match versus not being a good match. These quantities are then
+    used to rank specific addons for recommendation.
+
+    This will load a json file containing updated list of addon donors updated periodically
+    by a separate weekly process using Longitdudinal Telemetry data.
+
+    This recommender may provide useful recommendations when collaborative_recommender
+    may not work.
+    """
+
+    def __init__(self):
+        # Download the addon donors list.
+        self.donors_list = fetch_donors_list_remote_resource()
+
+        if self.donors_list is None:
+            logger.error("Cannot download the JSON resource: {}".format(DONOR_LIST_KEY))
+
+        # Download the probability mapping curves from similarity to likelihood of being a good donor.
+        self.lr_curves = fetch_lr_curve_remote_resource()
+
+        if self.lr_curves is None:
+            logger.error("Cannot download the JSON resource: {}".format(SIM_TO_P_FUN_KEY))
+
+    def can_recommend(self, client_data):
+        # We can't recommend if we don't have our data files.
+        if self.donors_list is None or self.lr_curves is None:
+            return False
+
+        # Check that the client info contains a non-None value for each required telemetry field.
+        required_fields = ["geoCity", "subsessionLength", "locale", "os", "bookmark_count", "tab_open_count",
+                           "total_uri", "unique_tlds"]
+
+        for single_required_field in required_fields:
+            client_value = client_data.get(single_required_field, None)
+            if client_value is None:
+                logger.error("Malformed client data encountered in SimilarityRecommender, {} field not detected"
+                             .format(single_required_field))
+                return False
+        return True
+
+    def get_lr(self, score):
+        # Compute the difference between each element in the lr curve score index and the current score.
+        index = [abs(score - s[0]) for s in self.lr_curves]
+
+        # Find the index of the closest value that was precomputed in lr_curves
+        _, idx = min((val, idx) for (idx, val) in enumerate(index))
+
+        numer_val = self.lr_curves[idx][1][0]
+        denum_val = self.lr_curves[idx][1][1]
+
+        # Compute LR based on numerator and denominator values
+        return float(numer_val) / float(denum_val)
+
+    def compute_similarity_to_donors(self, client_data):
```

Comment:
Let's add documentation for this function.

nit: maybe `get_similar_donors` is a more appropriate name here, since we're returning the candidate set of donors, right?

Reply:
What do you think of the suggested function name change here?

Reply:
oops, missed this one on the last pass. Now it is changed.